src/style/style.js

```diff
@@ -554,21 +557,26 @@ class Style extends Evented {
             return;
         }
 
-        if (typeof layerObject.source === 'object') {
-            this.addSource(id, layerObject.source);
-            layerObject = clone(layerObject);
-            layerObject = (extend(layerObject, {source: id}): any);
-        }
+        let layer;
+        if (layerObject.type === 'custom') {
+            layer = createStyleLayer(layerObject);
```

Comment:
Since this isn't going through the style spec validation path, the provided implementation needs to be validated 1) for layer id uniqueness, and 2) conformance to the CustomLayerInterface.  Probably 1) makes sense to do here, and maybe 2) in the CustomStyleLayer constructor?